feat: implement generic interface for plugins

frenchwr opened the pull request:
This PR is motivated by the request in https://github.com/snapcrafters/gimp/issues/447 for the upstream GIMP packaging team to take over ownership of this snap.

# What has changed

* Qt G'MIC part removed from snap and migrated to a [standalone snap](https://github.com/canonical/gimp-plugins-gmic-snap)
* OpenVINO plug removed from snap
* Generic plug introduced that supports connecting to multiple slots and mounts their content at `$SNAP/gimp-plugins`
* OpenVINO command chain scripts removed from snap
* Generic command chain script introduced that supports loading wrapper scripts from content snaps, which is helpful for running setup commands (e.g. exporting environment variables)
* OpenVINO plugin docs removed from README and snapcraft.yaml
* Generic plug docs added to README

# What has not changed

* Multiple plug definitions needed for supporting OpenVINO plugins. Unfortunately these cannot be delegated to the `openvino-ai-plugins-gimp` snap as GIMP itself needs access to the host resources or libs provided by these plugs.

# Testing

I have tested on my laptop with the following commands.

```
sudo snap install --dangerous ./gimp_3.0.4_amd64.snap 
sudo snap install openvino-toolkit-2404 --channel latest/edge
sudo snap install gimp-plugins-gmic --channel latest/edge
sudo snap connect gimp:gimp-plugins gimp-plugins-gmic:gimp-plugins
sudo snap install openvino-ai-plugins-gimp --channel latest/edge
sudo snap connect gimp:gimp-plugins openvino-ai-plugins-gimp:gimp-plugins
sudo snap connect gimp:dot-local-share-openvino-ai-plugins-gimp
```

Note the last command is only needed because the GIMP snap was built and installed locally. Normally this would auto-connect when installed from the Snap Store. We could also pursue auto-connections for the `gimp-plugins-gmic` and `openvino-ai-plugins-gimp` snaps.

* Screenshot showing OpenVINO's stable diffusion plugin in action:

<img width="1513" height="1306" alt="Screenshot from 2025-09-05 11-26-11" src="https://github.com/user-attachments/assets/dcfa1286-9df9-4eac-b03e-e80cdacc2455" />

* Screenshot showing Qt G'MIC in action, editing the image resulting from the stable diffusion plugin

<img width="1534" height="1872" alt="Screenshot from 2025-09-05 11-27-32" src="https://github.com/user-attachments/assets/36d6d589-29b5-43d5-ba1c-3bdae684d72b" />


# Related work

* [Qt G'MIC plugins snap](https://github.com/canonical/gimp-plugins-gmic-snap)
* https://github.com/canonical/openvino-toolkit-snap/pull/24
* https://github.com/canonical/openvino-ai-plugins-gimp-snap/pull/18

frenchwr committed:
feat: implement generic interface for plugins

Jehan commented:
Hi! I am not sure to understand. Did you not manage to have the OpenVino into its own plug-in snap? Only G'MIC? So that means the OpenVino plug-in is dropped?

brunvonlope commented:
@Jehan Yes. From my understanding of the patches OpenVino, GMIC (and any other 3P plug-in) will "mount" on `/PATH_TO_SNAP_FILES/gimp-plugins` like we mount on `/PATH_TO_FLATPAK_FILES/extensions`

Actually, it was mostly separated but the initialization of some OpenVino things like GPU etc was being done hardcoded in GIMP Snap files. Now, it is done by `command-chain/enable-plugins`.

And I request your security opinion about such enable-plugins Shell file because it allows to run anything the 3P plugin want before GIMP command is run?

brunvonlope commented:
@Jehan I ported this MR to our Nightly snap so you can understand the diff easier: https://gitlab.gnome.org/GNOME/gimp/-/merge_requests/2452/diffs

frenchwr committed:
fix: re-add plugs providing openvino and npu libs

frenchwr commented:
Please note I just added a small commit to add back a few plugs to the GIMP app section. I initially removed these when I began the work and forgot to re-add them. The plug definition was in the original PR but they were not actually being used.

frenchwr commented:
@brunvonlope Yes, I think that captures it well. 

> Did you not manage to have the OpenVino into its own plug-in snap? Only G'MIC? So that means the OpenVino plug-in is dropped?

OpenVINO is already in its own plug-in snap called `openvino-ai-plugins-gimp`. With this PR we no longer have direct references to this snap (except maybe the `dot-local-share` plug described below). It's been replaced by a generic interface that can be used by OpenVINO, G'MIC, and others in the future.

The only parts I was unable to remove were the following plugs:

* `intel-npu`: provides access to the Intel NPU device node(s) in `/dev/accel` on the host. The OpenVINO plugins need access to these nodes but since the plugins are run by the GIMP process you cannot delegate access to the `openvino-ai-plugins-gimp` snap.
* `npu-libs`: These are the .so's for the Intel NPU user mode driver and compiler. It's possible to build these into the `openvino-ai-plugins-gimp` snap but then you lose the advantages of running and testing different versions easily.
* `openvino-libs`: similar to `npu-libs`. These provide .so's and Python code used by the `openvino-ai-plugins-gimp` snap.
* `dot-local-share-openvino-ai-plugins-gimp`: grants r/w access to `$HOME/.local/share/openvino-ai-plugins-gimp` on the host for installing AI models used by the OpenVINO plugins. The `openvino-ai-plugins-gimp` snap distributes a command that can be used to install models to this same path, or you can install them directly through GIMP (see the model management window in my screenshot above).

frenchwr commented:
> And I request your security opinion about such enable-plugins Shell file because it allows to run anything the 3P plugin want before GIMP command is run?

I am definitely open to enhancing the script if security is a concern. I purposefully kept it simple to reduce maintenance and make it easy to understand for future plugin packagers.

brunvonlope commented:
@frenchwr `intel-npu` and `npu-libs` plugs are provided by which Snap? They seem to be generic so that wouldn't be a problem to maintain

Regarding `openvino-libs` and `dot-local-share-openvino-ai-plugins-gimp`, I tell you in advance that, while I understand the explanation, they can't go on in the official snap (if it comes to GIMP umbrella). As upstream, we follow a maintenance principle of never specificying plugin-specific (non-vanilla) things in packages. I mean, it is a good first step but still a lot of OpenVINO-specific things are in the .yml

frenchwr commented:
> @frenchwr intel-npu and npu-libs plugs are provided by which Snap? They seem to be generic so that wouldn't be a problem to maintain

They are both provided by the `intel-npu-driver` snap.

brunvonlope commented:
Yeah, I'm fine with these two from `intel-npu-driver`👍. While they are hardware-specific is not like we have a lot of CPU vendors around there to care of

frenchwr commented:
> Regarding openvino-libs and dot-local-share-openvino-ai-plugins-gimp, I tell you in advance that, while I understand the explanation, they can't go on in the official snap (if it comes to GIMP umbrella). As upstream, we follow a maintenance principle of never specificying plugin-specific (non-vanilla) things in packages. I mean, it is a good first step but still a lot of OpenVINO-specific things are in the .yml

I see. For `openvino-libs` I could build those directly into the `openvino-ai-plugins-gimp` snap. It'd be a bit of a bummer but it's something we can do.

For `dot-local-share-openvino-ai-plugins-gimp` is it just the name of the plug that is the issue? If so perhaps we could rename it and make it something more general purpose for the GIMP snap.

The other option is to only allow users to install models from GIMP, and not from this command that is distributed with the `openvino-ai-plugins-gimp` snap. Then we would not need a shared location and could simply install the models in `SNAP_USER_DATA` or `SNAP_USER_COMMON`.

brunvonlope commented:
> If so perhaps we could rename it and make it something more general purpose for the GIMP snap.

Would be nice. On GIMP we normally point user data to `$HOME/.config/GIMP/MAJOR.MINOR`

Jehan commented:
> And I request your security opinion about such enable-plugins Shell file because it allows to run anything the 3P plugin want before GIMP command is run?

Well sure, from a security standpoint, it does mean that these third-party snaps can run anything they want. Now I guess it's still within the main sandbox, and yeah when you add code from various sources, you do weaken security. You have to somehow trust the sources for all the code you run. It's not that different from any other system.

That's all the more reasons why we want this code separated. We don't want to be responsible for it. We will provide the "entry point" for third-party plug-ins to plug into the GIMP snap, but in the end, it's the people's job (the ones who install GIMP and these third-party plug-ins) to decide if they trust all the software source (us, but also the third-party plug-in developers).

So I don't see more of a problem than that.

Ideally though if the only use of this all is just to add some environment variables (at least, that's what I see in the [.env](https://github.com/canonical/gimp-plugins-gmic-snap/blob/main/wrapper/gmic.env) in the G'MIC snap, though I could not find a `.env` in the [OpenVino snap](https://github.com/canonical/openvino-ai-plugins-gimp-snap)… or is that supposed to be [this file](https://github.com/canonical/openvino-ai-plugins-gimp-snap/blob/main/command-chain/openvino-ai-plugins-gimp-launch)?), we could make up some more basic format only accepting `var=value` code.

But if it really needs to be random shell commands, then I guess… 🤷

jnsgruk commented:
Overall I like the approach. It still seems like we have more than we *might* need like `openvino-libs` as a plug - if we can avoid that then we should, but I also don't think it'll cause the upstream any issues if we can't find a workaround.

As for `dot-local-share-openvino-ai-plugins-gimp` - if we can make sure that the plugins just use a directory that GIMP already has access to, I think that would be preferable.

brunvonlope commented:
> but I also don't think it'll cause the upstream any issues if we can't find a workaround.

It will open a precedent to list plugin specific plugs etc in an upstream manifest. I see that for a downstreamer this is easy and fun but for us (upstreamers) that maintain official packages used by many people this is not acceptable. We can't maintain a manifest with plugin developers reporting to add plugin-specific plugs: etc to privilege specific plugins. We need to stay neutral.

frenchwr commented:
> Overall I like the approach. It still seems like we have more than we might need like openvino-libs as a plug - if we can avoid that then we should, but I also don't think it'll cause the upstream any issues if we can't find a workaround.

> As for dot-local-share-openvino-ai-plugins-gimp - if we can make sure that the plugins just use a directory that GIMP already has access to, I think that would be preferable.

Ok - let me work on removing these later (hopefully today). We'll lose some convenient features but nothing essential.

frenchwr commented:
> Ideally though if the only use of this all is just to add some environment variables (at least, that's what I see in the [.env](https://github.com/canonical/gimp-plugins-gmic-snap/blob/main/wrapper/gmic.env) in the G'MIC snap, though I could not find a .env in the [OpenVino snap](https://github.com/canonical/openvino-ai-plugins-gimp-snap)… or is that supposed to be [this file](https://github.com/canonical/openvino-ai-plugins-gimp-snap/blob/main/command-chain/openvino-ai-plugins-gimp-launch)?), we could make up some more basic format only accepting var=value code.

That's the one (well, a slight re-named variation of it is in a PR). It's mostly just setting environment vars but there is a python script that gets run and is needed to do device detection (i.e. CPU, GPU, NPU) at runtime.

frenchwr committed:
chore: remove remaining openvino plugs

frenchwr commented:
The `openvino-libs` and `dot-local-share-openvino-ai-plugins-gimp` plugs are now gone, and I've tested that the plugins still function correctly.

I managed this by (i) installing the `openvino-toolkit-2404` snap directly into the `openvino-ai-plugins-gimp` snap via `stage-snaps` and (ii) switching to a model installation path that is private to the GIMP snap at `~/snap/gimp/common` (also known as `SNAP_USER_COMMON`).

frenchwr committed:
chore: remove plug-declaration

jnsgruk commented:
Nice! Thanks Will! @Jehan & @brunvonlope this must be getting pretty close for you? :)

brunvonlope commented:
@jnsgruk Seems fine 👍. I will just wait for Jehan final word. I didn't totally understand what he said in this comment https://github.com/snapcrafters/gimp/pull/460#issuecomment-3259912476 about the wrapper concept:

https://github.com/snapcrafters/gimp/blob/4157064c5feaf4f04e46bdcfc4ea5356752c97c9/command-chain/enable-plugins#L9

Jehan commented:
@brunvonlope I mostly said that it's not ideal, but well… I was a bit fatalistic, but maybe we should strive for the best Snap interface. If we settle for "meh", we'll regret it later and it will be hard to change once all plug-ins will start using this.

> That's the one (well, a slight re-named variation of it is in a PR). It's mostly just setting environment vars but there is a python script that gets run and is needed to do device detection (i.e. CPU, GPU, NPU) at runtime.

Why isn't this device detection code in the plug-in itself? In fact, could someone tell me exactly when these `.env`  files are run? Is it when you start snapped GIMP, the snap runs all the `.env` files before it starts GIMP?

If it's to generate some data which needs to be in the Snap, couldn't it be run during the Snap creation?

And if it needs absolutely to be run at the start of GIMP, each time the plug-in is updated, this code can be run inside the `query_procedures()` implemention of the `GimpPlugIn`. Or if it really needs to be run at each and every startup, the plug-in developers can implement [`init_procedures()`](https://developer.gimp.org/api/3.0/libgimp/vfunc.PlugIn.init_procedures.html) instead.

frenchwr commented:
> Why isn't this device detection code in the plug-in itself? 

The issue is that the plugins rely on device detection at install-time ([you can see the logic here](https://github.com/intel/openvino-ai-plugins-gimp/blob/main/gimpopenvino/install_utils.py#L168-L186), if you're interested), which is not ideal as generally the snap build environment and snap runtime environment can be very different. I've raised this with the plugin developers but at the moment the snap overcomes this by re-running their logic for device detection each time GIMP is run.

> In fact, could someone tell me exactly when these `.env` files are run? Is it when you start snapped GIMP, the snap runs all the `.env` files before it starts GIMP?

Yes, exactly. Btw, I realize that file naming is a bit unusual. I was looking for a unique naming pattern and I thought `.env` at least conveyed (in part) their intention. However, I'm happy to change the pattern or even the design if there are concerns. I'm open to ideas. :)

frenchwr commented:
@Jehan @brunvonlope What do you think about the state of this work? Any other feedback or suggestions?

brunvonlope commented:
@Jehan Ping to see https://github.com/snapcrafters/gimp/pull/460#issuecomment-3268448947

Jehan commented:
I was planning on suggesting a MR to https://github.com/intel/openvino-ai-plugins-gimp/

I just haven't had the time to look at it yet. But anyway moving their hardware-detection code to `query_procedures()` (which will be run by GIMP every time the plug-in is updated) should be extra simple and would get rid of the whole issue of having to run it by the Snap.

My idea is that if we are lax now, then later it's too late because we have created the interface allowing plug-ins to just run random code out of GIMP control. And once the interface is there, it will be hard to remove.

P.S.: if anyone gets what I suggest — it should really not be hard to implement at all! —, feel free to make the MR to OpenVino plug-in developers. I don't need to be the one doing it and I've already zillions of things to do.

brunvonlope commented:
@frenchwr Now I think things are clearer. What is blocking this MR is the implementation of all the initialization in the plug-in itself. There is the Python reference of GIMP API: https://lazka.github.io/pgi-docs/#Gimp-3.0.

Then, without the wrapping concept, we can merge this MR.

Jehan commented:
If anyone wants to do the MRs, it's pretty easy. Search for `query_procedures` in the plug-ins: https://github.com/search?q=repo%3Aintel%2Fopenvino-ai-plugins-gimp%20query_procedures&type=code

On the OpenVino side, I see there are 3 such plug-ins. Then move the `install_utils` code (currently in `.env` file) in there.

Well it's probably not needed in all 3 plug-ins. According to the comments I see [in the `.env`](https://github.com/canonical/openvino-ai-plugins-gimp-snap/blob/main/command-chain/openvino-ai-plugins-gimp-launch), it looks like this is only for the super resolution and semseg plug-ins. But even running it twice seems too much. So either this can be split into 2 calls (in the semseg plug-in, you only generate the semseg models and in super-res plug-in only the relevant models). Or else maybe add some flag file with the date or something like this, so the first plug-in to run generates all the data and the second doesn't, after having verified the flag time (or some other way to make sure only one plug-in does the work if several are installed).

I was mentioning possibly switching to `init_procedures()` in a previous comment, but that's probably not needed. If the code is in `init_procedures()`, it will be run unconditionally at every startup. In `query_procedures()`, it will only be run every time you update the plug-in. From what I gathered, it should be enough. No need to overdo it.

The whole patch should take just a few minutes to make. That's the testing which is currently blocking me as it means I'd have to install the plug-in, test it, make it sure everything works fine how it should (while I never used it until now) with the patch, etc. It would not take hours, yet still more than someone who already has it all set up. So really anyone, feel free to make this patch before I do. 🤗

frenchwr commented:
Thanks for the suggestions, @Jehan . I'm open to proposing something to the maintainers, I expect they will be open to it but it may take some weeks for them to test, merge, release, etc. In the meantime we can maintain a patch in the snap package of the plugins. I will take a closer look later this week.

You hinted before that you are ok with plugin snaps exporting environment variables within a wrapper that gets called before running GIMP - just making sure that is still the case? If so, I can make changes in this PR that enforces this, where we accept only shell variables expressed as key-value pairs from the `.env` wrappers. Let me know if this sounds reasonable.

Jehan commented:
@frenchwr That would be a lesser evil, but is it absolutely needed even? I mean, this is not needed in system packages. This is not needed on Flatpak. Why would this be needed in Snaps?

GIMP plug-ins should just have their environment straight when GIMP calls them. It's not very complicated (and it works with the same plug-ins on any other system I know).

frenchwr commented:
> GIMP plug-ins should just have their environment straight when GIMP calls them. It's not very complicated (and it works with the same plug-ins on any other system I know).

Would you mind describing how this works with Flatpaks? For example, if a plugin ships a .so file how does GIMP find it?

Jehan commented:
Ok so I investigated a bit.

1/ So first of all, in our flatpak, we create an extension point for plug-ins in third-party flatpaks, defined [here](https://github.com/flathub/org.gimp.GIMP/blob/master/org.gimp.GIMP.json#L60-L67) in our flatpak manifest. What it says is that all plug-ins data will be installed in `/app/extensions/` and GIMP will indeed load all plug-ins installed in `/app/extensions/plug-ins/`.

2/ Now when I enter a shell inside the Flatpak sandbox (`flatpak run --devel --command=bash org.gimp.GIMP//stable`), I check where it finds the Qt libraries:

```sh
$ ldd /app/extensions/plug-ins/gmic_gimp_qt/gmic_gimp_qt  |grep Qt
	libQt5Widgets.so.5 => /app/extensions/GMic/lib/libQt5Widgets.so.5 (0x00007f5234200000)
	libQt5Gui.so.5 => /app/extensions/GMic/lib/libQt5Gui.so.5 (0x00007f5233a00000)
	libQt5Network.so.5 => /app/extensions/GMic/lib/libQt5Network.so.5 (0x00007f5233855000)
	libQt5Core.so.5 => /app/extensions/GMic/lib/libQt5Core.so.5 (0x00007f5232a00000)
```

It finds it inside /app/extensions/GMic/lib/ even though `LD_LIBRARY_PATH` is not edited.

3/ As it turns out, it's because it is simply added to `/etc/ld.so.conf` inside the flatpak through a conf file:

```sh
$ cat /run/flatpak/ld.so.conf.d/app-002-org.gimp.GIMP.Plugin.GMic.conf 
/app/extensions/GMic/lib
```

And that's thanks to this option in the extension point linked above (IIUC):

```
            "add-ld-path": "lib",
```

This is documented here: https://docs.flatpak.org/en/latest/extension.html

> `add-ld-path` adds a path relative to the extension prefix /app/extension_directory/foo/lib, so that a library residing there can be loaded. This supports a single value. If the $LD_LIBRARY_PATH environment variable exists, the path here is appended to it, otherwise it is put in a ld config file in /run/flatpak/ld.so.conf.d.

Ideally something similar could exist for snap extensions: we could tell them to just install their files some way and the Snap sandbox will be automatically updated to handle these.

Could you tell me how it's currently organized in the Snap by the way?

frenchwr commented:
Thanks for the detailed explanation, @Jehan .

> Could you tell me how it's currently organized in the Snap by the way?

I'm assuming you mean how snaps handle loading .so's from other snaps. The general approach (as far as I'm aware) is to use the [content interface](https://snapcraft.io/docs/content-interface) and update LD_LIBRARY_PATH, generally from a command-chain script (as we're doing for GIMP in this PR), which is just a wrapper that gets run before the snapped command (e.g. `gimp`). 

We _could_ do something similar to the Flatpak for finding .so's specifically...it would look similar to our current design where the GIMP snap expects to find custom path information in a certain directory within each of the plug-in snaps.

What about other environment variables (not just LD_LIBRARY_PATH) that need to get passed in to a plug-in snap? Is that supported by the Flatpak version? At least in the case of the OpenVINO plugins there are a fair number needed. For example, `PATH` needs an update to find a `optimum-cli` command for compiling ML models, and the plugin version is determined from git tags so the snap needs some shell variables to run git commands successfully. I realize these examples may not be using best practices, but I'm curious if exporting other environment variables is possible.

brunvonlope commented:
Ok. To add my two cents as the AppImage packager of GIMP, there are many environment variables (most I know, but may be some I don't know) that can interact with GTK apps in unpredictable ways and crash GIMP. Not to mention other dependencies that have environment variables like libthai, libheif, ghostscript, the list goes on and on.

So, from my experience, **I really suggest to drop this wrapper concept entirely** and make OpenVINO behave like "normal" plugins. I think this can be archieved with setenv() etc.

frenchwr commented:
Ok, understood. Let me see what I can do.

frenchwr commented:
I spent some time on this yesterday. For better or worse, using command-chain scripts that update environment variables is the standard method for a content consumer snap (GIMP in this case) to interact with a content producer snap (GIMP plugins in this case). 

There is no `add-ld-path` equivalent in the snap world. I could build some similar functionality for GIMP (and I started down this path yesterday) but ultimately there will still need to be a script that is managed by GIMP that processes paths pointing to .so's for various plugins. There is simply no way around it, unfortunately.

To add a bit more context, the `gnome` snapcraft extension (which is used [here](https://github.com/frenchwr/gimp/blob/candidate/snap/snapcraft.yaml#L75)) applies a few command-chain scripts of its own (you can think of a snapcraft extension as a sort of preprocessor directive that adds several sections to the `snapcraft.yaml` before the snap is built) which sets many environment variables (and runs other shell commands) prior to launching GIMP to ensure GTK and graphical support are provided. So the GIMP snap will continue  running command-chain scripts prior to running GIMP itself even if we were to somehow remove the command-chain script from generic plugins interface.

brunvonlope commented:
> There is simply no way around it, unfortunately.

Can't we, then, rework the `enable-plugins` command-chain to only add to LD_LIBRARY_PATH? I am perfectly fine if we manage to do this and seems to be possible.

> So the GIMP snap will continue running command-chain scripts prior to running GIMP itself even if we were to somehow remove the command-chain script from generic plugins interface.

Well, one thing is GNOME extension, which is a dependency to run GIMP (a GTK app), and a plug-in, which is not crucial to GIMP run. I'm aware and fine with all the GNOME extension is injecting.

frenchwr commented:
> Can't we, then, rework the enable-plugins command-chain to only add to LD_LIBRARY_PATH? I am perfectly fine if we manage to do this and seems to be possible.

Great! Let me try this.

> Well, one thing is GNOME extension, which is a dependency to run GIMP (a GTK app), and a plug-in, which is not crucial to GIMP run. I'm aware and fine with all the GNOME extension is injecting.

Ok good, glad you're aware. :)

frenchwr committed:
chore: only allow plug-ins to append to LD_LIBRARY_PATH

frenchwr commented:
@brunvonlope @Jehan I have just pushed an update that now only allows plugins to append to `LD_LIBRARY_PATH` through the `enable-plugins` command-chain script. It required some patches to the G'MIC and OpenVINO source to conform with this design, but I have now tested that everything works with this new design.

Let me know what you think!

brunvonlope commented:
@frenchwr LGTM ✅ Thanks for your hard work

Could you help me with this eror? https://gitlab.gnome.org/GNOME/gimp/-/jobs/5613598#L2958

I am trying to port the generic interface to the upstream snap but no success.

jnsgruk commented:
> Could you help me with this eror? https://gitlab.gnome.org/GNOME/gimp/-/jobs/5613598#L2958

I *think* you just need to make the `enable-plugins` file executable 🙂 

```
gimp bruno/snap-3p-plugins
 ❯ ls -la build/linux/snap/patches/
.rw-rw-r-- 617 jon jon  1 Oct 13:45 -- enable-plugins
```

brunvonlope commented:
I get different permissions when checking the file on Windows:

```
brunv@LAPTOP-BUBJBGPC CLANG64 /c/Users/brunv/source/repos/gimp/build/linux/snap/patches
$ls -l enable-plugins
-rwxr-xr-x 1 brunv brunv 639 Oct  1 08:19 enable-plugins
```

jnsgruk commented:
Interesting, could try just setting the permissions explicitly in the `snapcraft.yaml`, e.g.

```yaml
#...
command-chain-plugins:
    plugin: dump
    source: build/linux/snap/patches/
    override-pull: |
      craftctl default
      chmod +x enable-plugins
    organize:
      enable-plugins: command-chain/enable-plugins
```

I *think* that would do it - though I've not run it to test

brunvonlope commented:
Thanks. It worked 🎊

<img width="1283" height="800" alt="image" src="https://github.com/user-attachments/assets/2c90a736-ebad-4fa2-83f2-79ba3616b64d" />

---

Would you mind trying this snap https://gitlab.gnome.org/GNOME/gimp/-/jobs/5613801/artifacts/browse/build/linux/snap/_Output/ with the OpenVINO plugins in your machine just to be sure?

jnsgruk commented:
I'll wait for @frenchwr for that one, since I don't have an OpenVINO compatible machine at the moment! 🙂

jnsgruk commented:
One thing that we can do in the mean time is make sure you have an account setup at https://snapcraft.io as the "official" GIMP account - it needs to be set up with a group mailing list for the project (i.e. contact@gimp.org) - and then we can work on transferring the snap and making it a "Verified Publisher" account.

brunvonlope commented:
> One thing that we can do in the mean time is make sure you have an account setup at https://snapcraft.io/ as the "official" GIMP account - it needs to be set up with a group mailing list for the project (i.e. [contact@gimp.org](mailto:contact@gimp.org)) - and then we can work on transferring the snap and making it a "Verified Publisher" account.

@Jehan ping

Jehan commented:
> @brunvonlope @Jehan I have just pushed an update that now only allows plugins to append to `LD_LIBRARY_PATH` through the `enable-plugins` command-chain script. It required some patches to the G'MIC and OpenVINO source to conform with this design, but I have now tested that everything works with this new design.
> 
> Let me know what you think!

Looks very fine to me @frenchwr ! Simple and working. Perfect.

> One thing that we can do in the mean time is make sure you have an account setup at https://snapcraft.io as the "official" GIMP account - it needs to be set up with a group mailing list for the project (i.e. [contact@gimp.org](mailto:contact@gimp.org)) - and then we can work on transferring the snap and making it a "Verified Publisher" account.

I am making an alias dedicated to this package: https://github.com/gnome-infra/ansible/pull/31

The address will be snap at our domain. We'll tell you when it's up.

frenchwr committed:
chore: remove git dependency

frenchwr commented:
> Would you mind trying this snap https://gitlab.gnome.org/GNOME/gimp/-/jobs/5613801/artifacts/browse/build/linux/snap/_Output/ with the OpenVINO plugins in your machine just to be sure?

I have tested on my laptop and they work well together. There was one issue with the version string (related to using git inside the snap) which I have fixed by removing the git dependency from the OpenVINO plugin snap and from the GIMP snap as well.

One last thing I want to do is test on a system with a NPU. I will do that later today or tomorrow and report back here.

Jehan commented:
@jnsgruk @frenchwr Now that our special Snap alias has been created, I tried to create an account for the GIMP team. I logically named it with login "gimp" and the website tells me:

> gimp is already in use.

Could you please verify whoever is using this login has a good reason to do it? For info, we have also started some legal proceeding in the last few weeks to get rid of parasitic usage, people basically impersonating the team.

If the person has a real use case (maybe they are actually named Mr or Ms. Gimp; or maybe it's their actual personal nickname for years, they publish completely unrelated software and they are absolutely not trying to make it look like they are related to the GIMP team), it can be OK. I'm not asking to shut down other people's correct usage.

I'll wait until this is verified before making an account. Because if it's a correct usage, I'll use another login, like gimpteam or something. But if this looks like a parasitic account, I'll appreciate if it can be rid of and I'll make the account under the name `gimp`, as it seems the better name.

Jehan commented:
Hmmm… after looking at some snaps, and following links, I guess this is where we'd see if the login `gimp` has any snaps currently: https://snapcraft.io/publisher/gimp

It looks like they never published any snap. But maybe this login is not only for Snap publishers?…

jnsgruk commented:
> @jnsgruk @frenchwr Now that our special Snap alias has been created, I tried to create an account for the GIMP team. I logically named it with login "gimp" and the website tells me:
> 
> > gimp is already in use.
> 
> Could you please verify whoever is using this login has a good reason to do it? For info, we have also started some legal proceeding in the last few weeks to get rid of parasitic usage, people basically impersonating the team.
> 
> If the person has a real use case (maybe they are actually named Mr or Ms. Gimp; or maybe it's their actual personal nickname for years, they publish completely unrelated software and they are absolutely not trying to make it look like they are related to the GIMP team), it can be OK. I'm not asking to shut down other people's correct usage.
> 
> I'll wait until this is verified before making an account. Because if it's a correct usage, I'll use another login, like gimpteam or something. But if this looks like a parasitic account, I'll appreciate if it can be rid of and I'll make the account under the name `gimp`, as it seems the better name.

Let me get back to you. I know we have some "reserved" names in the Store to prevent people squatting so I'll look into it.

jnsgruk commented:
@Jehan Okay, try now! `gimp` should be available on the store/launchpad

Jehan commented:
LOL I had an old Ubuntu login which I tried earlier (to see if it still existed), and now I can't log as anything else! Despite me having logged out, each time I click "Sign in", it just relogs me. I obviously tried to delete all snapcraft.io cookies.

frenchwr committed:
fix: add back NPU dir to LD_LIBRARY_PATH

Jehan commented:
Oh nevermind. This was stored as a ubuntu.com cookie! Ugh. Small suggestion: when someone clicks "Log out", they mean it. So the ubuntu login should be logged out too! Or whatever other solution, like having some cookie saying that the snapcraft.io is logged out despite the central Ubuntu login is still ON. Or whatever. Otherwise how can someone have several accounts (personal account + team account for instance here!).

Anyway account "gimp" successfully created @jnsgruk !

frenchwr commented:
> One last thing I want to do is test on a system with a NPU. I will do that later today or tomorrow and report back here.

I'm glad I tested because I had inadvertently dropped $SNAP/npu-libs from the LD_LIBRARY_PATH.  This is fixed in https://github.com/snapcrafters/gimp/pull/460/commits/75eda115db09d85e14ddced8faccb3b1159b6b96. 

@brunvonlope Would you mind applying that in your gitlab build so I can test, please?

jnsgruk commented:
Okay, so now we just need to figure out the next steps. Now you have an account, I'll add you as a collaborator on the snap, then I can just make the forum post to get the snap formally transferred to you if you're ready for it?

Then there is another forum-based process to get you verified, which I can also kick off.

I assume you're registered on forum.snapcraft.io - if you could let me know your username I'll make sure to ping you in the posts.

brunvonlope commented:
> @brunvonlope Would you mind applying that in your gitlab build so I can test, please?

@frenchwr Thanks, added with https://gitlab.gnome.org/GNOME/gimp/-/merge_requests/2452/diffs?diff_id=1488332&start_sha=429de30aa619e70cd31d3de93ec1fb1506f1809d

There is the new artifacts link to test: https://gitlab.gnome.org/GNOME/gimp/-/jobs/5618825/artifacts/browse/build/linux/snap/_Output/

brunvonlope commented:
> Okay, so now we just need to figure out the next steps.

@jnsgruk I think we should first merge this MR (after the testing that @frenchwr is doing) then continue the discussion on the issue. This is easier to follow IMO

jnsgruk commented:
@brunvonlope No probs - just let me know when you're ready for that stuff 🙂

jnsgruk commented:
Also, seems like we're pretty close. If Will is happy that this now works, I think we should land it and get it out into the snap channels soon?

frenchwr committed:
fix: prepend instead of append to LD_LIBRARY_PATH

In case plugins ship .so's that are already present inside GIMP,
prefer the ones that come with the plugins. This puts the maintenance
responsibility on the plugin maintainer for ensuring compatability.

frenchwr commented:
@brunvonlope I have pushed another fix (see the commit message for details): https://github.com/snapcrafters/gimp/pull/460/commits/f777258903688807e689e3d7dc7ebff71bec2b02

I was seeing a crash on one of the NPU systems before deploying this fix.

I have now tested this on two different generations of NPU and everything looks good, but just with my local build. If you could generate a new gitlab build with this fix I will do one last (hopefully!) round of testing.

brunvonlope commented:
@frenchwr @Jehan I wonder if https://github.com/snapcrafters/gimp/pull/460/commits/f777258903688807e689e3d7dc7ebff71bec2b02 will not make GIMP load the libraries from the plug-in snap first. From my understanding RUNPATH don't have precedente over LD_LIBRARY_PATH.

So, this would conflict with the way flatpak behaves? https://github.com/snapcrafters/gimp/pull/460#issuecomment-3302174496

> If the $LD_LIBRARY_PATH environment variable exists, the path here is appended to it, otherwise it is put in a ld config file in /run/flatpak/ld.so.conf.d.

frenchwr commented:
> @frenchwr @Jehan I wonder if [f777258](https://github.com/snapcrafters/gimp/commit/f777258903688807e689e3d7dc7ebff71bec2b02) will not make GIMP load the libraries from the plug-in snap first. 

Yes, that was my intention, but as I've thought more about it I suppose this is too dangerous, as it means a plug-in could break GIMP if it introduces a conflicting lib that GIMP depends on.

The reason I need the change is that the OpenVINO plugin snap ships some bleeding edge graphics libs for supporting newer Intel hardware. The plugins do not run at all on Intel Lunar Lake CPUs (which have a GPU and NPU built-int) when using the stock graphical libs that ship with GIMP (actually these are delivered via a different content producer snap).

I will re-visit tomorrow. Most likely I will revert [f777258](https://github.com/snapcrafters/gimp/commit/f777258903688807e689e3d7dc7ebff71bec2b02) and manage LD_LIBRARY_PATH from the plugin code so that it picks up the newer graphics libs only from its process space without impacting GIMP itself.

frenchwr committed:
Revert "fix: prepend instead of append to LD_LIBRARY_PATH"

This reverts commit f777258903688807e689e3d7dc7ebff71bec2b02.

frenchwr commented:
Ok, I indeed reverted the last commit, so I'm back to appending to LD_LIBRARY_PATH instead of prepending. This took some refactoring work on the OpenVINO plugins (it's amazing how such a seemingly simple change can complicate things...) but I now have it working on the Lunar Lake system that gave me some problems before. I expect the other two systems should also work fine but let me test those (first thing Monday) just to be sure then we can merge.

Hope everyone has a nice weekend. :)

brunvonlope commented:
@frenchwr Did the last upstream .snap I linked (which is the same as yours after the revert) worked with OpenVINO?

frenchwr commented:
> @frenchwr Did the last upstream .snap I linked (which is the same as yours after the revert) worked with OpenVINO?

Yes, I just need to test on a few more machines on Monday.

frenchwr commented:
Success! I've now tested the GIMP 3.1.5 snap from the most recent Gitlab build on three different machines spanning different generations of hardware and Ubuntu releases: both the G'MIC and OpenVINO plugins work across the board.

I also rebased against the tip of `candidate` so https://github.com/snapcrafters/gimp/commit/d3169eeb15cc8c66f7749ccf20b2ef296c172ace is also now included in this PR.

jnsgruk commented:
Okay, great! I'll take one final look and merge this once the CI passes! Thank you, Will!

One thing we probably need to figure out is where to put the docs for adding the OpenVINO and GMIX plugins when using the Snap - perhaps there is somewhere in the upstream docs we can do that @Jehan?

brunvonlope commented:
> One thing we probably need to figure out is where to put the docs for adding the OpenVINO and GMIX plugins when using the Snap - perhaps there is somewhere in the upstream docs we can do that @Jehan?

The binaries will be stored in /snap. Are you talking about the gimp-plugins interface?

jnsgruk commented:
> Are you talking about the gimp-plugins interface?

Yeh, I'm talking about a brief docs entry we can send existing users of the snap to, so they know how to get GMIC back, or get OpenVINO back - where previously it was installed as part of the snap, they now need to add extra snaps that place content in the new plugins interface :)

brunvonlope commented:
@jnsgruk I think that, for now, we need to mention in the Snap Store description.

Ideally, Snap Store itself should do that. See: https://github.com/canonical/snapcraft.io/issues/5240 and https://github.com/ubuntu/app-center/issues/1954

jnsgruk commented:
Oh that's easy enough - just needs adding to the `description` field in your `snapcraft.yaml` and it'll get populated on the store :)

Jehan commented:
> One thing we probably need to figure out is where to put the docs for adding the OpenVINO and GMIX plugins when using the Snap - perhaps there is somewhere in the upstream docs we can do that @Jehan?

I don't think we want to favor specific plug-ins in the documentation. But we can put down command lines to install these in the news which we'll probably publish to announce our new Snap.

Ideally though, Snap should have an internal logic to "discover" extensions for another package. That's what they did in Flatpak so that on GNOME Software or other installer GUI, the "GIMP" page has a list of all extensions which can optionally be installed.

brunvonlope commented:
> Ideally though, Snap should have an internal logic to "discover" extensions for another package. That's what they did in Flatpak so that on GNOME Software or other installer GUI, the "GIMP" page has a list of all extensions which can optionally be installed.

@Jehan Yeah, I filled two issues for this. For the short term, we need to mention these two plugins to not disrupt greatly users expectations.

@jnsgruk But in the future I should set a ETA to remove the mentions. Then, we will need collaboration of Cannonical guys with the issues I linked.

Jehan commented:
Just to make it clearer, the reason why this should not be maintained manually is:

* On our side, we don't want to favor specific extensions, which also means that all the cool ones we never heard about get shadowed by the most famous which most already know anyway.
* An automatic system would allow people to write their own extensions without going through us! We don't even have to like or dislike an extension. We provide a public API and people are welcome to use it without asking us. Then they could upload on Snap Store and it should be discoverable by anyone without us having to "validate" the extension.

jnsgruk commented:
> Then, we will need collaboration of Cannonical guys with the issues I linked.

Yes, I can work on that - not immediately - but I can start some discussions on how we solve that problem going forward!

brunvonlope commented:
@Jehan Just to clarify too: I don't pretend to maintain manually new plugins in a list. What I mean is that we need to mention **just these two** historic plugins that were being shipped before the generic interface. Otherwise, will be mostly impossible to find them until it is not implemented on Snap Store frontend 😅
